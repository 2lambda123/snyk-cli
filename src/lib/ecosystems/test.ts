import * as snyk from '../index';
import * as config from '../config';
import { isCI } from '../is-ci';
import { makeRequest } from '../request/promise';
import { Options } from '../types';
// import { TestCommandResult } from '../../cli/commands/types';
import * as spinner from '../../lib/spinner';
import { EcosystemPlugin, ScanResult, TestResult } from './types';
import { TestDependenciesResponse } from '../snyk-test/legacy';
import { assembleQueryString } from '../snyk-test/common';

export async function testEcosystem(
  plugin: EcosystemPlugin,
  paths: string[],
  options: Options,
): Promise<Result[]> {
  // TODO: this is an intermediate step before consolidating ecosystem plugins
  // to accept flows that act differently in the testDependencies step
  // if (plugin.test) {
  //   const { readableResult: res } = await plugin.test(paths, options);
  //   return TestCommandResult.createHumanReadableTestCommandResult(res, '');
  // }
  const scanResultsByPath: { [dir: string]: ScanResult[] } = {};
  for (const path of paths) {
    await spinner(`Scanning dependencies in ${path}`);
    options.path = path;
    const pluginResponse = await plugin.scan(options);
    scanResultsByPath[path] = pluginResponse.scanResults;
  }
  spinner.clearAll();
  const allResults = await testDependencies(scanResultsByPath, options);
  // const { errors, testResults } = separateErrorResults(allResults);
  // const stringifiedData = JSON.stringify(testResults, null, 2);
  // if (options.json) {
  //   return TestCommandResult.createJsonTestCommandResult(stringifiedData);
  // }
  // const emptyResults: ScanResult[] = [];
  // const scanResults = emptyResults.concat(...Object.values(scanResultsByPath));
  // const readableResult = await plugin.display(
  //   scanResults,
  //   testResults,
  //   errors,
  //   options,
  // );

  // return TestCommandResult.createHumanReadableTestCommandResult(
  //   readableResult,
  //   stringifiedData,
  // );
  return allResults;
}

// function separateErrorResults(
//   results: Result[],
// ): {
//   errors: string[];
//   testResults: TestResult[];
// } {
//   const errors: string[] = [];
//   const testResults: TestResult[] = [];

//   for (const i of results) {
//     if ('error' in i) {
//       errors.push(i.error.message);
//     } else {
//       testResults.push(i.testResult);
//     }
//   }

//   return { errors, testResults };
// }
interface ResultSuccess {
  path: string;
  scanResult: ScanResult;
  testResult: TestResult;
}

interface ResultError {
  path: string;
  scanResult: ScanResult;
  error: Error;
}

export type Result = ResultError | ResultSuccess;

async function testDependencies(
  scans: {
    [dir: string]: ScanResult[];
  },
  options: Options,
): Promise<Result[]> {
  const results: Result[] = [];
  for (const [path, scanResults] of Object.entries(scans)) {
    await spinner(`Testing dependencies in ${path}`);
    for (const scanResult of scanResults) {
      const payload = {
        method: 'POST',
        url: `${config.API}/test-dependencies`,
        json: true,
        headers: {
          'x-is-ci': isCI(),
          authorization: 'token ' + snyk.api,
        },
        body: {
          scanResult,
        },
        qs: assembleQueryString(options),
      };
      try {
        const response = await makeRequest<TestDependenciesResponse>(payload);
        results.push({
          path,
          scanResult,
          testResult: {
            issues: response.result.issues,
            issuesData: response.result.issuesData,
            depGraphData: response.result.depGraphData,
          },
        });
      } catch (error) {
        if (error.code >= 400 && error.code < 500) {
          throw new Error(error.message);
        }
        results.push({
          path,
          scanResult,
          error: new Error('Could not test dependencies in ' + path),
        });
      }
    }
  }
  spinner.clearAll();
  return results;
}
