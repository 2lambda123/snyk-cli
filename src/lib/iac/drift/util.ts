import * as fs from 'fs';

export function createIfNotExists(path: string) {
  if (!fs.existsSync(path)) {
    fs.mkdirSync(path, { recursive: true });
  }
}
