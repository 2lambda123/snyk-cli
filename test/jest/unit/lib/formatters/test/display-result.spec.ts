import * as fs from 'fs';
import stripAnsi from 'strip-ansi';
import { displayResult } from '../../../../../../src/lib/formatters/test/display-result';
import { getWorkspacePath } from '../../../../util/getWorkspacePath';
import { getFixturePath } from '../../../../util/getFixturePath';

describe('displayResult', () => {
 
  it('Pip result with pins', () => {
    const withRemediation = JSON.parse(
      fs.readFileSync(
        getFixturePath('pip-app-with-remediation/test-graph-results.json'),
        'utf8',
      ),
    );

    const res = displayResult(
      withRemediation,
      { showVulnPaths: 'all', path: 'src' },
      3,
    );
    expect(stripAnsi(res).replace(/\[http.*\]/g, '[URL]')).toMatchSnapshot();
  });

  it('with license issues', () => {
    const withRemediation = JSON.parse(
      fs.readFileSync(
        getWorkspacePath(
          'ruby-app/test-graph-response-with-legal-instruction.json',
        ),
        'utf8',
      ),
    );

    const res = displayResult(
      withRemediation,
      { showVulnPaths: 'all', path: 'src' },
      3,
    );
    expect(stripAnsi(res).replace(/\[http.*\]/g, '[URL]')).toMatchSnapshot();
  });

  it('with Upgrades & Patches', () => {
    const withRemediation = JSON.parse(
      fs.readFileSync(
        getFixturePath(
          'npm-package-with-severity-override/test-graph-result-patches.json',
        ),
        'utf8',
      ),
    );

    const res = displayResult(
      withRemediation,
      { showVulnPaths: 'all', path: 'src' },
      3,
    );
    expect(stripAnsi(res).replace(/\[http.*\]/g, '[URL]')).toMatchSnapshot();
  });
});
