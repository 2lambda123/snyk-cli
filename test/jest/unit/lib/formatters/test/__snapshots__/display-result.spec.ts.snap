// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displayResult Docker test result 1`] = `
"
Testing src...

✗ High severity vulnerability found in musl
  Description: Out-of-bounds Write
  Info: [URL]
  Introduced through: meta-common-packages@meta
  From: meta-common-packages@meta > musl@1.1.19-r10
  Fixed in: 1.1.19-r11

✗ High severity vulnerability found in expat
  Description: XML External Entity (XXE) Injection
  Info: [URL]
  Introduced through: expat@2.2.5-r0, .python-rundeps@0, git@2.18.2-r0
  From: expat@2.2.5-r0
  From: .python-rundeps@0 > expat@2.2.5-r0
  From: git@2.18.2-r0 > expat@2.2.5-r0
  Remediation:
    Upgrade direct dependency expat@2.2.5-r0 to expat@2.2.7-r0 (triggers upgrades to expat@2.2.7-r0)
    Some paths have no direct dependency upgrade that can address this issue.
  Fixed in: 2.2.7-r0

✗ High severity vulnerability found in expat
  Description: XML External Entity (XXE) Injection
  Info: [URL]
  Introduced through: expat@2.2.5-r0, .python-rundeps@0, git@2.18.2-r0
  From: expat@2.2.5-r0
  From: .python-rundeps@0 > expat@2.2.5-r0
  From: git@2.18.2-r0 > expat@2.2.5-r0
  Remediation:
    Upgrade direct dependency expat@2.2.5-r0 to expat@2.2.7-r1 (triggers upgrades to expat@2.2.7-r1)
    Some paths have no direct dependency upgrade that can address this issue.
  Fixed in: 2.2.7-r1



Organization:      gitphill
Package manager:   undefined
Docker image:      src
Licenses:          enabled

Tested 40 dependencies for known issues, found 3 issues.

Tip: Detected multiple supported manifests (3), use --all-projects to scan all of them at once.

Snyk wasn’t able to auto detect the base image, use \`--file\` option to get base image remediation advice.
Example: $ snyk container test src --file=path/to/Dockerfile

To remove this message in the future, please run \`snyk config set disableSuggestions=true\`"
`;

exports[`displayResult Docker test result no file path and base image auto detected 1`] = `
"
Testing src...

✗ High severity vulnerability found in musl
  Description: Out-of-bounds Write
  Info: [URL]
  Introduced through: meta-common-packages@meta
  From: meta-common-packages@meta > musl@1.1.19-r10
  Fixed in: 1.1.19-r11

✗ High severity vulnerability found in expat
  Description: XML External Entity (XXE) Injection
  Info: [URL]
  Introduced through: expat@2.2.5-r0, .python-rundeps@0, git@2.18.2-r0
  From: expat@2.2.5-r0
  From: .python-rundeps@0 > expat@2.2.5-r0
  From: git@2.18.2-r0 > expat@2.2.5-r0
  Remediation:
    Upgrade direct dependency expat@2.2.5-r0 to expat@2.2.7-r0 (triggers upgrades to expat@2.2.7-r0)
    Some paths have no direct dependency upgrade that can address this issue.
  Fixed in: 2.2.7-r0

✗ High severity vulnerability found in expat
  Description: XML External Entity (XXE) Injection
  Info: [URL]
  Introduced through: expat@2.2.5-r0, .python-rundeps@0, git@2.18.2-r0
  From: expat@2.2.5-r0
  From: .python-rundeps@0 > expat@2.2.5-r0
  From: git@2.18.2-r0 > expat@2.2.5-r0
  Remediation:
    Upgrade direct dependency expat@2.2.5-r0 to expat@2.2.7-r1 (triggers upgrades to expat@2.2.7-r1)
    Some paths have no direct dependency upgrade that can address this issue.
  Fixed in: 2.2.7-r1



Organization:      gitphill
Package manager:   undefined
Docker image:      src
Base image:        base Image
Licenses:          enabled

Tested 40 dependencies for known issues, found 3 issues.

Tip: Detected multiple supported manifests (3), use --all-projects to scan all of them at once."
`;

exports[`displayResult Docker test result with base image name not found warning 1`] = `
"
Testing alpine:latest...

✗ High severity vulnerability found in musl
  Description: Out-of-bounds Write
  Info: [URL]
  Introduced through: meta-common-packages@meta
  From: meta-common-packages@meta > musl@1.1.19-r10
  Fixed in: 1.1.19-r11



Organization:      test-organization
Package manager:   apk
Target file:       Dockerfile
Project name:      alpine
Docker image:      alpine:latest
Platform:          linux/amd64
Licenses:          enabled

Tested 40 dependencies for known issues, found 1 issue.

Tip: Detected multiple supported manifests (1), use --all-projects to scan all of them at once.

Warning: Unable to analyse Dockerfile provided through \`--file\`.
         Dockerfile must begin with a FROM instruction. This may be after parser directives, comments, and globally scoped ARGs.

Pro tip: use \`--exclude-base-image-vulns\` to exclude from display Docker base image vulnerabilities.

To remove this message in the future, please run \`snyk config set disableSuggestions=true\`"
`;

exports[`displayResult Docker test result with base image non resolvable warning 1`] = `
"
Testing alpine:latest...

✗ High severity vulnerability found in musl
  Description: Out-of-bounds Write
  Info: [URL]
  Introduced through: meta-common-packages@meta
  From: meta-common-packages@meta > musl@1.1.19-r10
  Fixed in: 1.1.19-r11



Organization:      test-organization
Package manager:   apk
Target file:       Dockerfile
Project name:      alpine
Docker image:      alpine:latest
Platform:          linux/amd64
Licenses:          enabled

Tested 40 dependencies for known issues, found 1 issue.

Tip: Detected multiple supported manifests (1), use --all-projects to scan all of them at once.

Warning: Unable to analyse Dockerfile provided through \`--file\`.
         Dockerfile must have default values for all ARG instructions.

Pro tip: use \`--exclude-base-image-vulns\` to exclude from display Docker base image vulnerabilities.

To remove this message in the future, please run \`snyk config set disableSuggestions=true\`"
`;

exports[`displayResult Docker test result with remediation advice 1`] = `
"
Testing src...

✗ High severity vulnerability found in musl
  Description: Out-of-bounds Write
  Info: [URL]
  Introduced through: meta-common-packages@meta
  From: meta-common-packages@meta > musl@1.1.19-r10
  Fixed in: 1.1.19-r11

✗ High severity vulnerability found in expat
  Description: XML External Entity (XXE) Injection
  Info: [URL]
  Introduced through: expat@2.2.5-r0, .python-rundeps@0, git@2.18.2-r0
  From: expat@2.2.5-r0
  From: .python-rundeps@0 > expat@2.2.5-r0
  From: git@2.18.2-r0 > expat@2.2.5-r0
  Remediation:
    Upgrade direct dependency expat@2.2.5-r0 to expat@2.2.7-r0 (triggers upgrades to expat@2.2.7-r0)
    Some paths have no direct dependency upgrade that can address this issue.
  Fixed in: 2.2.7-r0

✗ High severity vulnerability found in expat
  Description: XML External Entity (XXE) Injection
  Info: [URL]
  Introduced through: expat@2.2.5-r0, .python-rundeps@0, git@2.18.2-r0
  From: expat@2.2.5-r0
  From: .python-rundeps@0 > expat@2.2.5-r0
  From: git@2.18.2-r0 > expat@2.2.5-r0
  Remediation:
    Upgrade direct dependency expat@2.2.5-r0 to expat@2.2.7-r1 (triggers upgrades to expat@2.2.7-r1)
    Some paths have no direct dependency upgrade that can address this issue.
  Fixed in: 2.2.7-r1



Organization:      gitphill
Package manager:   undefined
Docker image:      src
Base image:        ubuntu:20.04
Licenses:          enabled

Tested 40 dependencies for known issues, found 3 issues.

Tip: Detected multiple supported manifests (3), use --all-projects to scan all of them at once.

Base Image    Vulnerabilities  Severity
ubuntu:20.04  19               0 critical, 0 high, 3 medium, 16 low

Recommendations for base image upgrade:

Major upgrades
Base Image              Vulnerabilities  Severity
ubuntu:impish-20210928  13               0 critical, 0 high, 2 medium, 11 low


Learn more: [URL]"
`;

exports[`displayResult Pip result with pins 1`] = `
"
Testing src...

Tested 2 dependencies for known issues, found 32 issues, 2 vulnerable paths.

Issues with no direct upgrade or patch (1 issue):

  ✗ [Low] Directory Traversal
    Info: https://security.snyk.io/vuln/SNYK-PYTHON-DJANGO-1066259
    Introduced through: django@1.6.1
    Introduced by:
    django@1.6.1
    Fixed in: 2.2.18, 3.0.12, 3.1.6



Issues to fix by upgrading dependencies (1 issue):

  Upgrade django@1.6.1 to django@2.2.18 to fix 1 issue

  ✗ [Medium] Content Spoofing
    Info: https://security.snyk.io/vuln/SNYK-PYTHON-DJANGO-72888
    Introduced through: django@1.6.1
    Introduced by:
    django@1.6.1



Organization:      lili
Package manager:   pip
Target file:       requirements.txt
Project name:      pip-app
Open source:       no
Project path:      src
Licenses:          enabled

Test summary:

  1 issues with no upgrade or patch
  1 fixable issues: 1 Medium

Tip: Try \`snyk fix\` to address these issues.\`snyk fix\` is a new CLI command in that aims to automatically apply the recommended updates for supported ecosystems.
See documentation on how to enable this beta feature: https://docs.snyk.io/snyk-cli/fix-vulnerabilities-from-the-cli/automatic-remediation-with-snyk-fix#enabling-snyk-fix

Tip: Detected multiple supported manifests (3), use --all-projects to scan all of them at once."
`;

exports[`displayResult with Upgrades & Patches 1`] = `
"
Testing src...

Tested 2 dependencies for known vulnerabilities, found 4 vulnerabilities, 4 vulnerable paths.
Patchable issues (1 issue):

  Patch available for node-uuid@1.4.0

  ✗ [Low (originally Medium)] Insecure Randomness
    Info: https://security.snyk.io/vuln/npm:node-uuid:20160328
    Introduced through: node-uuid@1.4.0
    Introduced by:
    node-uuid@1.4.0


Issues to fix by upgrading (3 issues):

  Upgrade qs@0.0.6 to qs@6.0.4 to fix 3 issues

  ✗ [Low (originally High)] Prototype Override Protection Bypass
    Info: https://security.snyk.io/vuln/npm:qs:20170213
    Introduced through: qs@0.0.6
    Introduced by:
    qs@0.0.6

  ✗ [Low (originally High)] Denial of Service (DoS)
    Info: https://security.snyk.io/vuln/npm:qs:20140806
    Introduced through: qs@0.0.6
    Introduced by:
    qs@0.0.6

  ✗ [Low (originally Medium)] Denial of Service (DoS)
    Info: https://security.snyk.io/vuln/npm:qs:20140806-1
    Introduced through: qs@0.0.6
    Introduced by:
    qs@0.0.6



Organization:      another-org
Package manager:   npm
Target file:       package-lock.json
Project name:      shallow-goof
Open source:       no
Project path:      src

Test summary:

  4 fixable issues: 4 Low

Tip: Detected multiple supported manifests (3), use --all-projects to scan all of them at once."
`;

exports[`displayResult with license issues 1`] = `
"
Testing src...

Tested 3 dependencies for known issues, found 6 issues, 8 vulnerable paths.
License issues (1 issue):

  ✗ [High] Unknown license
    Info: https://security.snyk.io/vuln/snyk:lic:rubygems:rack-cache:Unknown
    Introduced through: rack-cache@1.1
    Introduced by:
    rack-cache@1.1
    Legal instructions:
    ○ for LGPL-3.0 license: I am legal license instruction


Issues to fix by upgrading (7 issues):

  Upgrade rack@1.6.5 to rack@1.6.11 to fix 1 issue

  ✗ [Medium] Cross-site Scripting (XSS)
    Info: https://security.snyk.io/vuln/SNYK-RUBY-RACK-72567
    Introduced through: rack@1.6.5
    Introduced by:
    rack@1.6.5
    rack-cache@1.1 > rack@1.6.5
    rack-protection@1.5.3 > rack@1.6.5

  Upgrade rack-cache@1.1 to rack-cache@1.3.0 to fix 2 issues

  ✗ [Medium] Cross-site Scripting (XSS)
    Info: https://security.snyk.io/vuln/SNYK-RUBY-RACK-72567
    Introduced through: rack@1.6.5
    Introduced by:
    rack@1.6.5
    rack-cache@1.1 > rack@1.6.5
    rack-protection@1.5.3 > rack@1.6.5

  ✗ [High] HTTP Header Caching Weakness
    Info: https://security.snyk.io/vuln/SNYK-RUBY-RACKCACHE-20031
    Introduced through: rack-cache@1.1
    Introduced by:
    rack-cache@1.1

  Upgrade rack-protection@1.5.3 to rack-protection@2.0.0 to fix 4 issues

  ✗ [Low] Side-channel attack
    Info: https://security.snyk.io/vuln/SNYK-RUBY-RACKPROTECTION-20394
    Introduced through: rack-protection@1.5.3
    Introduced by:
    rack-protection@1.5.3

  ✗ [Medium] Timing Attack
    Info: https://security.snyk.io/vuln/SNYK-RUBY-RACKPROTECTION-20395
    Introduced through: rack-protection@1.5.3
    Introduced by:
    rack-protection@1.5.3

  ✗ [Medium] Directory Traversal
    Info: https://security.snyk.io/vuln/SNYK-RUBY-RACKPROTECTION-22019
    Introduced through: rack-protection@1.5.3
    Introduced by:
    rack-protection@1.5.3

  ✗ [Medium] Cross-site Scripting (XSS)
    Info: https://security.snyk.io/vuln/SNYK-RUBY-RACK-72567
    Introduced through: rack@1.6.5
    Introduced by:
    rack@1.6.5
    rack-cache@1.1 > rack@1.6.5
    rack-protection@1.5.3 > rack@1.6.5



Organization:      lili2311
Package manager:   rubygems
Open source:       no
Project path:      src
Licenses:          enabled

Test summary:

  1 license issues: 1 High
  7 fixable issues: 1 High, 5 Medium, 1 Low

Tip: Detected multiple supported manifests (3), use --all-projects to scan all of them at once."
`;
